### ISSUE TITLE
[toast] accessibility: Undo action can't be reached in time with a keyboard or screen reader, or its timing adjusted

### ISSUE THREAD

**Issue by grace-snow** [mui/base-ui#4253]
# Bug report

## Current behavior

The "Undo action" Toast example triggers an important Toast with an "undo" button inside it. **Keyboard and screen reader users can't undo the action because focus doesn't move to the toast notification**. 

This should be high priority to fix as it potentially blocks keyboard and screen reader users from being able to complete important actions. 

Potential WCAG failures:
* See [Understanding WCAG 2.2 Success Criterion 2.1.1 Keyboard](https://www.w3.org/WAI/WCAG21/Understanding/keyboard.html)
* See [Understanding WCAG 2.2 Success Criterion 2.2.1 Timing Adjustable](https://www.w3.org/WAI/WCAG21/Understanding/timing-adjustable.html)

## Expected behavior

Keyboard focus either needs to move to these Toast notifications or the toast needs to be the next tab stop, and the toast should persist until dismissed or focus moves back out of them.

## Reproducible example

To replicate

1. Navigate to [Toast: Undo action example](https://base-ui.com/react/components/toast#undo-action)
2. Using only a keyboard, activate the "perform action" button
3. Try to tab to the toast and activate the "Undo" button
4. Note the toast has been loaded at the end of the DOM and disappears before it can be easily reached 
5. Repeat the test using a screen reader with keyboard

## Base UI version

v1.2.0

## Which browser are you using?

Chrome

## Which OS are you using?

Mac OS

## Which assistive tech are you using (if applicable)?

Keyboard, VoiceOver

## Additional context

In this case, moving focus to the toast and back to the originating control on close (or near to it, if the originating control no longer exists) would be a good solution. The toast should remain until it has been intentionally dismissed.

**Comment by aarongarciah** [comment]
Hey @grace-snow, thanks for opening the issue.

> Keyboard and screen reader users can't undo the action because focus doesn't move to the toast notification.

> Keyboard focus either needs to move to these Toast notifications 

Moving the focus automatically to the Toast is not an option, as it'd disrupt the user's workflow. In this regard, a Toast is close to the [Alert pattern](https://www.w3.org/WAI/ARIA/apg/patterns/alert/) from APG, which states:

> "intended to provide important and potentially time-sensitive information without interfering with the user's ability to continue working, it is crucial they do not affect keyboard focus."

([source](https://www.w3.org/WAI/ARIA/apg/patterns/alert/#:~:text=Because%20alerts%20are%20intended%20to%20provide%20important%20and%20potentially%20time%2Dsensitive%20information%20without%20interfering%20with%20the%20user%27s%20ability%20to%20continue%20working%2C%20it%20is%20crucial%20they%20do%20not%20affect%20keyboard%20focus.%20The%20Alert%20Dialog%20Pattern%20is%20designed%20for%20situations%20where%20interrupting%20work%20flow%20is%20necessary.))

Keyboard users can use `F6` to jump to the region containing the toasts, although it's true that most users won't be familiar with this shortcut. This is a difficult problem to solve.

Screen reader users have means of reaching the toasts using the rotor.

> the toast needs to be the next tab stop and the toast should persist until dismissed or focus moves back out of them.

Toasts are very context-specific, and the docs are not trying to cover every potential use case, as it'd be impossible. Luckily, the behavior can be tweaked depending on the needs:
- The [`container`](https://base-ui.com/react/components/toast#ToastPortal-container) prop on the `Toast.Portal` component can be used to  change where toasts are placed in the DOM.
- The [`timeout`](https://base-ui.com/react/components/toast#ToastProvider-timeout) prop on the `Toast.Provider` (or the `timeout` option when calling `toastManager.add()`) can be used to set a custom timeout or to disable auto-dismissal (setting `timeout` to `0`).

I've opened a PR to increase the timeout of the undo demo from 5 seconds to 10 seconds: https://github.com/mui/base-ui/pull/4975. But again, toasts a very specific to each project and it's expected that this is tweaked per use case.

**Comment by grace-snow** [comment]
I recommend updating the documentation to outline the accessibility considerations when using this component. Toasts in particular fall foul of the [2.2.1 Timing Adjustable (Level A)](https://www.w3.org/WAI/WCAG21/Understanding/timing-adjustable.html) requirement, as it must be the end user who is able to adjust timings, not just the developer building an application. Noting these accessibility requirements and some recommended approaches to meet them in the docs will help people implement this component in a more accessible way. 

For example, suggested solutions could be: 
- Adding a setting on a site/application that allows people to extend notification timings or disable auto-dismiss
- moving focus to toasts _if_ they are important and include a dismiss/close button
- including on-screen messages instead of a toast or after a toast has disappeared
